SYCL/ESIMD/api/functional/type_coverage.hpp

```diff
@@ -0,0 +1,49 @@
+//===-- logger.hpp - Declarate functions and containers with that will be used
+//      for run tests with different types. -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the declaration of the function that let iterate over
+/// provided types and provide each type to the call call-operator from the
+/// provided class/struct.
+///
+//===----------------------------------------------------------------------===//
+
+namespace type_coverage {
+
+// Type pack to store types and underlying data type names to use with
+// type_name_string
+template <typename... T> struct named_type_pack {
+  const std::string names[sizeof...(T)];
+
+  template <typename... nameListT>
+  named_type_pack(nameListT &&... nameList)
+      : names{std::forward<nameListT>(nameList)...} {}
+};
+
+// Run action for each of types given by named_type_pack instance
+template <template <typename, typename...> class action,
+          typename... actionArgsT, typename... types, typename... argsT>
+inline void for_all_types(const named_type_pack<types...> &typeList,
+                          argsT &&... args) {
+
+  // run action for each type from types... parameter pack
+  int packExpansion[] = {
+      (action<types, actionArgsT...>{}(std::forward<argsT>(args)...),
+
+       0 // Dummy initialization value
+       )...};
```

Comment:
```suggestion
  bool succeed{true};
  
  // run action for each type from types... parameter pack
  int packExpansion[] = {
      (succeed &= action<types, actionArgsT...>{}(std::forward<argsT>(args)...),

       0 // Dummy initialization value
       )...};
  // .......
  return succeed;
```

What do you think about something like this? We might incapsulate the `passed` logic into the return values as well.

Reply:
I've done it in this commit: ebaa8fc5c3830529804a100c915b4e571b57a5d3